src/invoice/invoice.rs

```diff
@@ -0,0 +1,1166 @@
+        let signature = RecoverableSignature::from_compact(&signature_bytes[0..64], recovery_id)
+            .expect("Invalid signature bytes");
+        Ok(InvoiceSignature(signature))
+    }
+}
+
+impl CkbInvoice {
+    fn hrp_part(&self) -> String {
+        format!(
+            "ln{}{}{}",
+            self.currency.to_string(),
+            self.amount
+                .map_or_else(|| "".to_string(), |x| x.to_string()),
+            self.prefix
+                .map_or_else(|| "".to_string(), |x| x.to_string()),
+        )
+    }
+
+    // Use the lostless compression algorithm to compress the invoice data.
+    // To make sure the final encoded invoice address is shorter
+    fn data_part(&self) -> Vec<u5> {
+        let invoice_data = RawInvoiceData::from(self.data.clone());
+        let compressed = ar_encompress(invoice_data.as_slice()).unwrap();
+        let mut base32 = Vec::with_capacity(compressed.len());
+        compressed.write_base32(&mut base32).unwrap();
+        base32
+    }
+
+    pub fn is_signed(&self) -> bool {
+        self.signature.is_some()
+    }
+
+    fn hash(&self) -> [u8; 32] {
+        let hrp = self.hrp_part();
+        let data = self.data_part();
+        let preimage = construct_invoice_preimage(hrp.as_bytes(), &data);
+        let mut hash: [u8; 32] = Default::default();
+        hash.copy_from_slice(&Sha256::hash(&preimage).to_byte_array());
+        hash
+    }
+
+    /// Recovers the public key used for signing the invoice from the recoverable signature.
+    pub fn recover_payee_pub_key(&self) -> Result<PublicKey, secp256k1::Error> {
+        let hash = Message::from_slice(&self.hash()[..])
+            .expect("Hash is 32 bytes long, same as MESSAGE_SIZE");
+
+        let res = secp256k1::Secp256k1::new()
+            .recover_ecdsa(&hash, &self.signature.as_ref().unwrap().0)
+            .unwrap();
+        Ok(res)
+    }
+
+    pub fn payee_pub_key(&self) -> Option<&PublicKey> {
+        self.data
+            .attrs
+            .iter()
+            .filter_map(|attr| match attr {
+                Attribute::PayeePublicKey(pk) => Some(pk),
+                _ => None,
+            })
+            .next()
+    }
+
+    pub fn payment_hash(&self) -> &[u8; 32] {
+        &self.data.payment_hash
+    }
+
+    pub fn payment_preimage(&self) -> Option<&[u8; 32]> {
+        self.data
+            .attrs
+            .iter()
+            .filter_map(|attr| match attr {
+                Attribute::PaymentPreimage(preimage) => Some(preimage),
+                _ => None,
+            })
+            .next()
+    }
+
+    pub fn payment_hash_id(&self) -> String {
+        hex::encode(self.payment_hash())
+    }
+
+    /// Checks if the signature is valid for the included payee public key or if none exists if it's
+    /// valid for the recovered signature (which should always be true?).
+    fn validate_signature(&self) -> bool {
```

Comment:
The name of this function seems to imply signature exist and is valid. Below we return true immediately if signature.is_none. From a cursory glimpse, I think all the `validate_signature` and `check_signature` logic should be moved into the `InvoiceBuilder` struct. In this way, we can just disallow creating `CkbInvoice` with invalid signature. BTW, https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/ is an interesting read.

Reply:
I had the same plan, but the validation also need `hash()` from invoice, 
```rust
   let hash = Message::from_slice(&self.hash()[..])
            .expect("Hash is 32 bytes long, same as MESSAGE_SIZE");
```

in `rust-lightning`, every invoice need to be signed by a private_key, and they build `RawBolt11Invoice` and then `SignedRawBolt11Invoice`, and then `Bolt11Invoice`, 
with a lot of wrapper attribute access methods, I think it's not a great way.